inorbit_edge/tests/test_robot_session.py

```diff
@@ -73,3 +76,79 @@ def test_robot_session_connect(monkeypatch):
             qos=1,
             retain=True,
         )
+
+
+def test_robot_session_custom_command_callback(monkeypatch):
+    def my_custom_command_handler(robot_session, msg):
+        pass
+
+    robot_session = RobotSession(
+        robot_id="id_123",
+        robot_name="name_123",
+        api_key="apikey_123",
+        custom_command_callback=my_custom_command_handler,
+    )
+
+    with requests_mock.Mocker() as mock:
+        mock.post(INORBIT_CLOUD_SDK_ROBOT_CONFIG_URL, json=ROBOT_CONFIG_MOCK_RESPONSE)
+        # mock mqtt client
+        mqtt_client_mock = MagicMock()
+        mqtt_client_mock.subscribe.return_value = (0, 0)
+        monkeypatch.setattr(robot_session, "client", mqtt_client_mock)
+        # mock publish and is_published so it always returns True
+        is_published_mock = MagicMock()
+        is_published_mock.is_published.return_value = True
+        publish_mock = MagicMock(return_value=is_published_mock)
+        monkeypatch.setattr(robot_session, "publish", publish_mock)
+        # connect robot_session so it populates properties with API response data
+        robot_session.connect()
+        # manually execute on_connect callback so the ``custom_command_callback``
+        # callback gets registered
+        robot_session._on_connect(..., ..., ..., 0)
+
+        assert robot_session.custom_command_callback == my_custom_command_handler
+        mqtt_client_mock.subscribe.assert_called_with(topic="r/id_123/custom_command")
+
+
+@pytest.mark.parametrize(
+    "msg_payload, log_string",
+    [
+        (b"\x8f\xac\x93\x87 \x8fG\x95v", "Failed to decode message, ignoring."),
+        (b'{"msg":"foobar"}}}}}}}', "Failed to parse JSON message, ignoring."),
+    ],
+)
+def test_robot_session_on_message_callback(caplog, msg_payload, log_string):
+    def my_custom_command_handler(robot_session, msg):
+        pass
+
+    with caplog.at_level(logging.ERROR, logger=RobotSession.__name__):
+        robot_session = RobotSession(
+            robot_id="id_123",
+            robot_name="name_123",
+            api_key="apikey_123",
+            custom_command_callback=my_custom_command_handler,
+        )
+
+        mqtt_message = MQTTMessage(topic=b"r/id_123/custom_command")
+        mqtt_message.payload = msg_payload
+
+        robot_session._on_message(..., ..., mqtt_message)
+        assert log_string in caplog.text
```

Comment:
instead of relying on the log, I'd verify that `my_custom_command_handler` was called using a flag set by that function or another mechanism.

Reply:
In this case the callback won't be called because it fails before when parsing/decoding the message